File: simple-iterator.py
#
# Copyright (c) 2022 Dirk Engel, https://skriptschmiede.de
#
# NAME
#     simple-iterator.py - Demo script to explain the
#     iterator pattern in Python
# SYNOPSIS
#     python simple-iterator.py
# DESCRIPTION
#     Simple iterator class example:
#     (1) __init__ method sets start values.
#     (2) __next__ method is customized to run our iteration.
#     (3) __iter__ method returns same object.
#     (4) End of iteration raises StopIteration.
# LICENSE
#     This file is licensed under the MIT License.
#     License text available at https://opensource.org/licenses/MIT
#


class TypingIterator:
    def __init__(self, my_container):
        """
        Example works for sequencial container objects.
        Pre-condition: Length and index of my_container
        must be defined from [0] to [-1].
        """

        self.index = 0
        self.my_container = my_container

    def __iter__(self):
        """
        We already have the __next__ method,
        so we return 'self' (same object) here.
        """
        return self

    def __next__(self):
        """
        Method __next__ contains the whole iterator logic.
        You could do any resumable calculation here.
        """
        # protocol requires: last element => StopIteration.
        if self.index + 1 > len(self.my_container):
            raise StopIteration
        # customized iteration values: types added
        value = (
            type(self.my_container[self.index]).__name__,
            self.my_container[self.index],
        )
        self.index += 1

        return value


if __name__ == "__main__":

    container_to_be_iterated = [{2, 8}, 9, "8"]

    # for-in loop over our simple iterator TypedIterator.
    print("\nOur 'typing' iterator:")
    for element in TypingIterator(container_to_be_iterated):
        print(repr(element))

File: simple-iterator.php
<?php

/*
Copyright (c) 2022 Dirk Engel, https://skriptschmiede.de

NAME
    simple-iterator.php - Demo script to explain the
    iterator pattern in PHP
SYNOPSIS
    php simple-iterator.php
DESCRIPTION
    Simple iterator class example:
    (1) __construct method is the iterator's constructor.
    (2) rewind method: go back to the iterator's initial state.
    (3) current method: return current iteration value.
    (4) key method: return current key.
    (5) next method: increment position.
    (6) valid method: check validity of current position.
LICENSE
    This file is licensed under the MIT License.
    License text available at https://opensource.org/licenses/MIT
 */

class TypingIterator implements Iterator
{
    private $position;
    private $my_container;

    public function __construct(array $my_container = [])
    {
        $this->position = 0;
        $this->my_container = $my_container;
    }

    public function rewind(): void
    {
        $this->position = 0;
    }

    public function current(): mixed
    {
        $item = $this->my_container[$this->position];
        return [gettype($item), $item];
    }

    public function key(): int
    {
        return $this->position;
    }

    public function next(): void
    {
        ++$this->position;
    }

    public function valid(): bool
    {
        return isset($this->my_container[$this->position]);
    }
}

$cont_to_be_iterated = new TypingIterator([[true, 7.7], 9, 'X']);

echo "\nOur 'typing' generator/iterator:\n";
foreach ($cont_to_be_iterated as $element) {
    echo $element[0], "\t\t", json_encode($element[1]), "\n";
}
